File: front/lib/domain/sync/entity_type.dart
import 'package:json_annotation/json_annotation.dart';

/// Closed enumeration of domain entity types eligible to the offline-first sync
/// protocol. Mirrors `persistence.model.EntityType` on the back.
enum EntityType {
  @JsonValue('ProductType')
  productType,
  @JsonValue('Organization')
  organization,
  @JsonValue('ProducerAccount')
  producerAccount,
  @JsonValue('Member')
  member,
  @JsonValue('MemberJoinRequest')
  memberJoinRequest,
  @JsonValue('Contract')
  contract,
  @JsonValue('DeliveryTemplate')
  deliveryTemplate,
  @JsonValue('OrganizationRequest')
  organizationRequest,
  @JsonValue('ProducerRequest')
  producerRequest,
  @JsonValue('Owner')
  owner,
  @JsonValue('MemberInvitation')
  memberInvitation,
  @JsonValue('OwnerInvitation')
  ownerInvitation,
  @JsonValue('BasketExchange')
  basketExchange,
  @JsonValue('Notification')
  notification,
  @JsonValue('DeviceToken')
  deviceToken,
  @JsonValue('AttendanceEmailRequest')
  attendanceEmailRequest,
  @JsonValue('ErrorReport')
  errorReport,
}

/// Wire-format mapping kept in sync with the `@JsonValue` annotations above.
/// Used by hand-rolled `MutationOp.Delete` JSON encoding (json_serializable's
/// generated map is private per file).
const Map<EntityType, String> entityTypeWireNames = {
  EntityType.productType: 'ProductType',
  EntityType.organization: 'Organization',
  EntityType.producerAccount: 'ProducerAccount',
  EntityType.member: 'Member',
  EntityType.memberJoinRequest: 'MemberJoinRequest',
  EntityType.contract: 'Contract',
  EntityType.deliveryTemplate: 'DeliveryTemplate',
  EntityType.organizationRequest: 'OrganizationRequest',
  EntityType.producerRequest: 'ProducerRequest',
  EntityType.owner: 'Owner',
  EntityType.memberInvitation: 'MemberInvitation',
  EntityType.ownerInvitation: 'OwnerInvitation',
  EntityType.basketExchange: 'BasketExchange',
  EntityType.notification: 'Notification',
  EntityType.deviceToken: 'DeviceToken',
  EntityType.attendanceEmailRequest: 'AttendanceEmailRequest',
  EntityType.errorReport: 'ErrorReport',
};

EntityType entityTypeFromWire(String value) => entityTypeWireNames.entries
    .firstWhere(
      (e) => e.value == value,
      orElse: () => throw FormatException('Unknown EntityType: $value'),
    )
    .key;

File: back/persistence/model/src/main/kotlin/EntityType.kt
package persistence.model

import kotlinx.serialization.Serializable

/**
 * Closed enumeration of domain entity types eligible to the offline-first sync
 * protocol. Acts as the discriminator for [EntityPayload], for [Change.entityType],
 * and as the key type for `cursors` / `changes` / `data` maps in `SyncRequest` /
 * `SyncResponse`.
 *
 * Adding a new synced entity is a deliberate four-step move:
 *  1. add a constant here,
 *  2. add a sealed subclass of [EntityPayload] that wraps the entity,
 *  3. add a sealed subclass of `EntityTypeService` to handle upsert/delete for it,
 *  4. extend the front contract.
 */
@Serializable
enum class EntityType {
    ProductType,
    Organization,
    ProducerAccount,
    Member,
    MemberJoinRequest,
    MemberInvitation,
    Contract,
    DeliveryTemplate,
    OrganizationRequest,
    ProducerRequest,
    Owner,
    OwnerInvitation,
    BasketExchange,
    Notification,
    DeviceToken,
    AttendanceEmailRequest,
    Producer,
    ErrorReport,
}

/** All synced entity types (union). */
val entityTypes: List<EntityType> = EntityType.entries
